Basic logging tests for cardano-wallet serve and launch

piotr-iohk opened the pull request:
# Issue Number

#350

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have added tests for cardano-wallet serve
- [ ] Will add tests for cardano-wallet launch also


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Commit:
Basic logging tests for cardano-wallet serve

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -857,3 +859,14 @@ postTransactionViaCLI ctx passphrase args = do
+-- | Get process out and err streams after n seconds of running and terminate
+getProcStream :: CreateProcess -> Int -> IO (String, String)
+getProcStream pr n = withCreateProcess pr $ \_ (Just stdout) (Just stderr) h -> do
+       threadDelay $ n*oneSecond
```
We should avoid delays as much as possible for these kind of tests because it creates very flaky tests in the end. It depends on many things in the host environment, and we often are too short in the delay. Instead, I'd suggest to follow a similar strategy to what we did for the file watching: 

- Wait for some event to appear
- Timeout the whole waiting time with a rather big upper-bound, like a minute or so.

So here, we could wait until we've seen in the log stream some lines, and wait at most 60s. If after 60s we haven't seen anything, then the test has failed.

piotr-iohk replied:
>So here, we could wait until we've seen in the log stream some lines, and wait at most 60s. If after 60s we haven't seen anything, then the test has failed.

Gotcha, however this wouldn't work for `--quiet` as it does not log anything in normal circumstances (ofc except for errors)

KtorZ replied:
Ah! Well, for the `--quiet`, maybe having a slightly bigger timeout (like 10s) to really leave time for some logs to appear?

piotr-iohk replied:
Hm, well `--quiet` logs only `error` so nothing is basically logged when everything is fine... perhaps one way is producing some error artificially to have something in the log (¯\_(ツ)_/¯) or just leave it as is for this particular case...

KtorZ replied:
What I mean is that, if it doesn't log anything after one seconde, we can't be sure that it has even started. If it doesn't log anything after 10, we still aren't sure, but at least it increases a bit our confidence.

KtorZ commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Server.hs:
```diff
@@ -30,8 +36,39 @@ spec = do
+            let process = proc' "cardano-wallet" args
+            (o, _) <- getProcStream process 1
+            o `shouldContain` "Debug"
+            o `shouldContain` "Warning"
```
I am not sure we should see any Warning :sweat_smile:  .... I mean, we do, and that's bad but I've  talked with the node-shell team already about lowering down the severity for the iohk-monitoring-framework statistics to debug or trace. It doesn't make sense to log those as "Warnings"

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -434,6 +436,34 @@ expectPathEventuallyExist filepath = do
             False ->
                 threadDelay (oneSecond `div` 2) >> doesPathExistNow
 
+-- | A file is eventually created on the given location
```
:point_up: sounds like a non-updated copy/paste

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -434,6 +436,34 @@ expectPathEventuallyExist filepath = do
+           waitForIt stdout h
+     where
+         waitForIt stdout h = do
+             is <- TIO.hGetLine stdout
```
I don't think GHC runtime will be happy if we try to `hGetLine` on an empty handle. There's maybe some wild IO exception error handling to do here :s

piotr-iohk replied:
I've added some handling there...

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -434,6 +436,34 @@ expectPathEventuallyExist filepath = do
+             case (wants `T.isInfixOf` is) of
+                 True -> do
+                     terminateProcess h
+                     return ()
```
suggestion: `void $ terminateProcess h`

piotr-iohk replied:
actually just `terminateProcess h` is enough...

piotr-iohk commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -434,6 +436,34 @@ expectPathEventuallyExist filepath = do
+                     terminateProcess h
+                     return ()
+                 False -> do
+                     threadDelay (oneSecond `div` 2)
```
this is probably not needed...

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -434,6 +436,34 @@ expectPathEventuallyExist filepath = do
+    oneSecond = 1_000_000
+
+    runProcUntil :: CreateProcess -> Text -> IO ()
+    runProcUntil pr wants = withCreateProcess pr $ \_ (Just stdout) _ h -> do
```
One minor thing but there a partial pattern-match here `Just stdout` which will fail hard with a non-helpful error if it fails. We only have a `Just` handle if the process was created with `CreatePipe` option, though, here, since we receive the `CreateProcess` as an argument, there's no way to _know_ that. I bet the call-site did it right, but, it'll be safer to either, overwrite the stdout stream parameter to be `CreatePipe` before calling `withCreateProcess`, or, to leave the creation of the process to the expect function.

piotr-iohk replied:
Ok, went for: `overwrite the stdout stream parameter to be CreatePipe`

Commit:
review oneSecond usage in DSL

Commit:
Make expectProcStdOutHas stop after looking at n lines of stdOut. I've changed it since `race` wasn't killing `serve` or `launch` when finding expected text on stdOut was not successful. Now it seems to be more reliable when failing (at least it should kill serve or launch and not affect subsequent tests).

There is one problem with this approach. When there is not enough lines
on the stdout for the given check to be performed the method
expectProcStdOutHas will wait until there is enough lines specified by
n.

Commit:
Remove checking for 'Warning' since it's no longer in the log it seems

KtorZ committed:
review style & indentation in newly added DSL functions

KtorZ committed:
replace expectProcessStdOutHas with 'collectStreams' and decouple assertions from logs gathering

KtorZ committed:
use 'collectStreams' in Server scenarios & LauncherSpec

This way, we avoid starting multiple times the process and make the tests a bit more resilient and faster
to run.

KtorZ committed:
re-generate nix machinery

KtorZ reviewed (approved):
Reviewed how logs get aggregated in order to avoid spawning processes 4 or 5 times in a single test scenario :s ... 
Expectations are now separated from the logs gathering which makes also things a bit easier to use / re-use.

cc @piotr-iohk

KtorZ committed:
also add logging test for Jörmungandr

KtorZ committed:
fix oversight: missing type-application in newly introduced integration test
